### PR TITLE
log newly added construction and runtime stats at randomly selected iterations

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51394 log newly added construction and runtime stats at randomly selected iterations**
* #51386 add stats that can only be collected at runtime

log newly added construction and runtime stats at randomly selected iterations

Differential Revision: [D26161885](https://our.internmc.facebook.com/intern/diff/D26161885/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26161885/)!

### CODE REVIEW

File: torch/lib/c10d/reducer.cpp [hunk 1]
```diff
@@ -988,6 +988,12 @@ void Reducer::prepare_for_forward() {
   std::lock_guard<std::mutex> lock(mutex_);
   num_iterations_++;
   forward_start_time_ = current_time_in_nanos();
+  // Log runtime (e.g. avg performance) stats at randomly selected iterations.
```
Comment (SciPioneer) [inline torch/lib/c10d/reducer.cpp:991]: "10, 1000, or 10000" are not randomly selected iterations and can cause confusion. Could you please update the comment for the consistency?

**Review by wayi1 — APPROVED** [review]

File: torch/lib/c10d/reducer.cpp (hunk 1 above)
Comment (zhaojuanmao) [reply torch/lib/c10d/reducer.cpp:991]: sounds good!

File: torch/lib/c10d/reducer.cpp [hunk 2]
```diff
@@ -988,6 +988,12 @@ void Reducer::prepare_for_forward() {
...
+  if (num_iterations_ == 10
```
Comment (rohan-varma) [inline torch/lib/c10d/reducer.cpp:992]: Can we insert these iterations into a constant set so that they can be easily modified and extended as needed?

Comment (zhaojuanmao) [reply torch/lib/c10d/reducer.cpp:992]: sounds good

**Pull request closed without merging**